experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -89,112 +93,117 @@ LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
+              branchOp.getConditionOperand());
+      inheritBB(specOp, branchDiscardNonSpec);
+
+      // The replicated branch directs the control token based on the path the
+      // speculative token took
+      auto branchReplicated = builder.create<handshake::ConditionalBranchOp>(
+          branchDiscardNonSpec->getLoc(),
+          branchDiscardNonSpec.getTrueResult() /* condition */,
+          ctrlSignal /* data */);
+      inheritBB(specOp, branchReplicated);
+
+      // Update ctrlSignal
+      ctrlSignal = branchReplicated->getResult(branchDirection);
+    }
     // Connect commit to the correct control signal and end traversal
     commitOp.setOperand(1, ctrlSignal);
   } else if (auto branchOp = dyn_cast<handshake::ConditionalBranchOp>(currOp)) {
-    // Replicate a branch in the control path and use new control signal.
-    // To do so, a structure of two connected branches is created.
-    // A speculating branch first discards the condition in case that
-    // the data is not speculative. In case it is speculative, a new branch
-    // is created that replicates the current branch.
-
-    MLIRContext *ctx = &getContext();
-    OpBuilder builder(ctx);
-    builder.setInsertionPointAfterValue(ctrlSignal);
-
-    // The speculating branch will discard the branch's condition token if the
-    // branch output is non-speculative. Speculative tag of the token is
-    // currently implicit, so the branch input itself is used at the IR level.
-    auto branchDiscardNonSpec = builder.create<handshake::SpeculatingBranchOp>(
-        branchOp.getLoc(), currOpOperand.get() /* specTag */,
-        branchOp.getConditionOperand());
-    inheritBB(specOp, branchDiscardNonSpec);
-
-    // The replicated branch directs the control token based on the path the
-    // speculative token took
-    auto branchReplicated = builder.create<handshake::ConditionalBranchOp>(
-        branchDiscardNonSpec->getLoc(),
-        branchDiscardNonSpec.getTrueResult() /* condition */,
-        ctrlSignal /* data */);
-    inheritBB(specOp, branchReplicated);
-
     // Follow the two branch results with a different control signal
     for (unsigned i = 0; i <= 1; ++i) {
       for (OpOperand &dstOpOperand : branchOp->getResult(i).getUses()) {
-        Value ctrl = branchReplicated->getResult(i);
-        routeCommitControl(markedPath, ctrl, dstOpOperand);
+        // Push the current branch information to the vector
+        // The items are referenced when the traversal hits a commit unit to
+        // build the commit control network.
+        branchTrace.emplace_back(currOpOperand.get(), branchOp, i);
+
+        // Continue traversal with new branchTracingList
+        routeCommitControlRecursive(ctx, specOp, arrived, dstOpOperand,
+                                    branchTrace);
+
+        // Pop the current branch information from the vector
+        // as it is no longer needed
+        branchTrace.pop_back();
```

Comment:
Should the pushing and polling be inside the for loop? Surely we can push and pop once for each branch

Reply:
That's correct, thanks!